### PR TITLE
feat(registration-block): editable success state

### LINKED ISSUE #1768: Registration Block: customisable post-registration state
After a user registers using the Registration Block, a text message replaces the registration UI. This state should be customisable (using the block editor, not just text).

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Because this code introduces a `save` handler on the block, a warning will be shown in the editor. Since the relevant features (Reader Activation) are behind a flag and not officially released yet, I don't think this is an issue. 

Due to constraints of the block editor – a block can have only one block-editable area ("inner blocks") – there's only one success state. This means that the same message (with note about confirmation email) will be displayed regardless of whether the reader has an account or not. This is accurate, though, because the email will be sent anyway. Also, this way the response does not reveal if the email exists in the database. WDYT @miguelpeixe ? 

I'm not aware of a convention for editing block "states" – here's the switch implemented in this PR:

image

Please share your thoughts @thomasguillot 👆

Closes #1768.

### How to test the changes in this Pull Request:

1. On `master`, insert the Reader Registration block on a page
2. Switch to this branch, observe the block working as before on front-end 
3. Open the editor, observe a notice about needing block recovery – click the block recovery button and observe the recovery succeeding 
4. Save the post, observe it's still working as expected on front-end 
5. In the editor, observe a new sidebar panel – "Edited state" – with a "Initial"/"Success" toggle
6. Edit the success state by switching the edited state
7. On the front-end, observe the new success state is displayed after registering 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by adekbadek** [comment]
I've noticed that there are actually two success states – "check your email" and "thank you for registering", but WP blocks allow only one inner-blocks container.

**Comment by miguelpeixe** [comment]
I think this could be handled universally through settings placed on the Engagement wizard

**Comment by miguelpeixe** [comment]
> Also, this way the response does not reveal if the email exists in the database.

We shouldn't be too worried about that. Anyone attempting to register a new account will know if the email exists in the database.

**Comment by thomasguillot** [comment]
We could use something similar to the [ToggleGroupControl](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-togglegroupcontrol--default) and have the states in the toolbar instead of in the sidebar.

Screenshot 2022-07-19 at 08 47 41

**Comment by adekbadek** [comment]
@thomasguillot – that's how I've initially implemented it, but the quirk of "inner blocks blocks" is that the top-level-block toolbar is not visible when there's content inside. For example with columns block, clicking on the block selects the child block, and its toolbar is visible.

**Comment by adekbadek** [comment]
@miguelpeixe –

> I think this could be handled universally through settings placed on the Engagement wizard (https://github.com/Automattic/newspack-plugin/pull/1773)

Yeah, the default value could be editable there.

**Comment by thomasguillot** [comment]
> that's how I've initially implemented it, but the quirk of "inner blocks blocks" is that the top-level-block toolbar is not visible when there's content inside

Oh I see. A bit annoying.

Ok, so I wonder if we could display some sort of "inner toolbar" at the top when the block (and child blocks) are selected.

Screenshot 2022-07-19 at 11 14 15
Screenshot 2022-07-19 at 11 17 24

**Comment by adekbadek** [comment]
@thomasguillot Good idea, don in f67274d ! Feels a bit weird with the toggle (suggesting an on/off state, not A/B switch), but gets the job done.

**Comment by adekbadek** [comment]
> Could we switch to buttons?

Done in d8e5818dd06df69297bf89842d95f93fa5bdee54

> Also, can we have this "toolbar" visible only on focus?

Someday, maybe, but I don't think it's a blocker for now.

**Comment by thomasguillot** [comment]
Pushed some small tweaks https://github.com/Automattic/newspack-plugin/pull/1785/commits/84c9862a2c7f2cd01b96f2c5993075507b40c69f

File: assets/blocks/reader-registration/edit.js [hunk 1]
```diff
@@ -11,14 +11,31 @@ import { intersection } from 'lodash';
...
+	PanelBody,
+	Button,
+} from '@wordpress/components';
 
 /**
  * Internal dependencies
  */
 import './editor.scss';
 
+const variantOptions = [
```
Comment (miguelpeixe) [inline assets/blocks/reader-registration/edit.js:35]: NIT: `variant` does not feel like a proper name for this. How about `blockState`?

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
The rendered block does not currently reflect the editor:

| Editor | Rendered |
| --- | --- |
| image |  image |
| image | image |

The text color and position of privacy text are off. Also, the initial state of the success message shows as a separate paragraph but renders in one line.

File: assets/blocks/reader-registration/edit.js (hunk 1 above)
Comment (adekbadek) [reply assets/blocks/reader-registration/edit.js]: Fixed in 2ae772000da5724b4a4589e2ee1f8132e7be2d2b

**Comment by adekbadek** [comment]
> The text color and position of privacy text are off. Also, the initial state of the success message shows as a separate paragraph but renders in one line.

Fixed in 2ae772000da5724b4a4589e2ee1f8132e7be2d2b & a5714b5d28bca8137e36d1851f50fda225008a31

**Review by miguelpeixe — APPROVED** [review]
🎉

**Pull request merged**